### PR TITLE
Remove jQuery and add tests to theme-a11y.js

### PR DESCRIPTION
**Pull request by t-kelly**

### CODE REVIEW

**Comment by huguestennier** [comment]
This is 🔥 🔥 

There's one problem I encountered recently, we need to find a way to only select the visible focusable elements.

Given this code : 
```HTML
<div id="container">
  <input id="textInput1" type="text" />
  <input id="textInput2" type="text" />
  <button id="button">Some button</button>
  <div style="opacity: 0;">
    <a href="#" id="link">hidden link</a>
  </div>
</div>
```
Our test `focuses the first tabable element after tabbing the last tabable element in a container` will fail, because if you tab 4 times, the test will receive `#link` and not `#textInput1`.

This can happen a lot when dealing with mega nav where you have second or third level navigation hidden. [Demo store example here](https://hugues-test-store.myshopify.com/?preview_theme_id=32166543416).

I had this fixed in the jQuery version by using [.filter(':visible')](https://api.jquery.com/visible-selector/) - maybe we can adapt it for our needs ? [Code here](https://github.com/jquery/jquery/blob/2d4f53416e5f74fa98e0c1d66b6f3c285a12f0ce/src/css/hiddenVisibleSelectors.js)

File: packages/theme-a11y/theme-a11y.puppeteer.test.js [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+      <div id="withoutDraggable"></div>
+    `);
+
+    const valid = ["draggableTrue"];
+    const invalid = ["draggableFalse", "withoutDraggable"];
+
+    await expectElements(valid, invalid);
+  });
+
+  test("selects only elements which are visible", async () => {
```
Comment (t-kelly) [inline packages/theme-a11y/theme-a11y.puppeteer.test.js]: @huguestennier does this test cover what's needed for visibility?

Comment (huguestennier) [reply packages/theme-a11y/theme-a11y.puppeteer.test.js]: Yup! Even tested when the parent is hidden but contains focusable elements like this : 

```HTML
<div style="display: none;">
  <button>Focusable but hidden button</button>
  <button>Focusable but hidden button</button>
</div>
```
and everything is ✅

File: packages/theme-a11y/theme-a11y.puppeteer.test.js [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    await expect(activeElement).toBe("textInput1");
+  });
+
+  test("focuses on the last tabable and visible element when shift-tabbing from first tabbable element", async () => {
+    await page.evaluate(() => {
+      const container = document.getElementById("container");
+      window.trapFocus(container);
+    });
+
+    await page.keyboard.down("Shift");
```
Comment (huguestennier) [inline packages/theme-a11y/theme-a11y.puppeteer.test.js]: Same thing here.

- focus the first focusable element
- Press shift + tab once

File: packages/theme-a11y/theme-a11y.puppeteer.test.js [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+      var ids = [];
+      window.focusable(document).forEach(element => ids.push(element.id));
+      return ids;
+    });
+
+    expect(matchIds).toEqual(expect.arrayContaining(valid));
+    expect(matchIds).not.toEqual(expect.arrayContaining(invalid));
+  }
+
+  test("is a function exported by a11y.js", () => {
```
Comment (huguestennier) [inline packages/theme-a11y/theme-a11y.puppeteer.test.js]: `theme-a11y.js`

File: packages/theme-a11y/theme-a11y.puppeteer.test.js [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+    await expectElements(valid, invalid);
+  });
+
+  test("selects valid <area> elements", async () => {
+    await page.setContent(`
+      <script type="module">
+        window.focusable = ${focusable.toString()};
+      </script>
+      <a id="withHref" href="#title"></a>
```
Comment (huguestennier) [inline packages/theme-a11y/theme-a11y.puppeteer.test.js]: Duplicated test with above, replace with `<area>`

File: packages/theme-a11y/theme-a11y.puppeteer.test.js [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+    const valid = ["visibleButton", "transparentButton"];
+    const invalid = ["hiddenButton", "hiddenChildButton"];
+
+    await expectElements(valid, invalid);
+  });
+});
+
+describe("trapFocus()", () => {
+  test("is a function exported by a11y.js", () => {
```
Comment (huguestennier) [inline packages/theme-a11y/theme-a11y.puppeteer.test.js]: `theme-a11y.js`

File: packages/theme-a11y/theme-a11y.puppeteer.test.js [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    expect(typeof trapFocus).toBe("function");
+  });
+});
+
+describe("removeTrapFocus()", () => {
+  test("is a function exported by a11y.js", () => {
```
Comment (huguestennier) [inline packages/theme-a11y/theme-a11y.puppeteer.test.js]: `theme-a11y.js`

File: packages/theme-a11y/theme-a11y.test.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  });
+});
+
+describe("bindInPageLinks()", () => {
+  beforeEach(() => {
+    document.body.innerHTML =
+      '<a id="link" href="#title"></a>' + '<h1 id="title">Title</h1>';
+  });
+
+  test("is a function exported by a11y.js", () => {
```
Comment (huguestennier) [inline packages/theme-a11y/theme-a11y.test.js]: `theme-a11y.js`

File: packages/theme-a11y/theme-a11y.puppeteer.test.js [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+    await expectElements(valid, invalid);
+  });
+
+  test("selects elements with draggable attribute", async () => {
+    await page.setContent(`
+      <script type="module">
+        window.focusable = ${focusable.toString()};
+      </script>
+      <div id="draggableTrue" draggable="true"></div>
+      <div id="draggableFalse" draggable="true"></div>
```
Comment (tauthomas01) [inline packages/theme-a11y/theme-a11y.puppeteer.test.js]: Is `draggable` supposed to be `false`?

File: packages/theme-a11y/theme-a11y.test.js [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+  test("adds an event handler that focuses the element referred to in an <a> element w/ a href='#...' when it is clicked", () => {
+    const link = document.getElementById("link");
+    const title = document.getElementById("title");
+
+    bindInPageLinks();
+
+    link.click();
+    expect(document.activeElement).toBe(title);
+  });
```
Comment (tauthomas01) [inline packages/theme-a11y/theme-a11y.test.js]: It should there should be a test for the opposite behaviour :

```
beforeEach(() => {
    document.body.innerHTML = 
    `
        <a id="link" href="#title"></a>
        <a id="regularLink" href="/test">Regular link</a>
        <h1 id="title">Title</h1>
    `;
});

test("does not focus the element referred to an <a> element without a href='#...' when it is clicked", () => {
    const regularLink = document.getElementById("regularLink");
    const title = document.getElementById("title");

    bindInPageLinks();

    link.click();
    expect(document.activeElement).toBe(regularLink);
});
```

Comment (t-kelly) [reply packages/theme-a11y/theme-a11y.test.js]: Difficult to test that behaviour in JSDOM. Going to just test green-path for now.

**Comment by t-kelly** [comment]
@huguestennier @tauthomas01 👍 ?

**Comment by huguestennier** [comment]
LGTM 👍

**Pull request merged**